### ISSUE TITLE
Provide a special class for pipeline-defined error messages

### ISSUE THREAD

**Issue by jfy133** [nextflow-io/nextflow#4937]
## New feature

It would be nice to have a special error class or function for pipeline-level errors or warnings, that help distinguish pipeline-defined (conditional) errors versus Nextflow level errors

## Usage scenario 

Pipelines will often have their own checks or validations, for example mutually exclusive parameters, that they will want to report to a user. Currently it is expected that a pipeline uses `error("error message")` to report this.

Often these are 'simple' errors, compared to Nextflow errors that have a more technical reason for the problem.

It might nice to be able to distinguish such pipeline-level errors from Nextflow errors. One benefit would be that it makes it much faster to identify where the error is coming from for developers, so they can quickly guide a user to a solution (i.e., whether it's a simple fix, or something the developer needs to investigate deeper.

Currently I do this manually by adding a prefix to warnings or logs, but it could be nice to have this happen automatically. Furthermore the way to report an error vs warning is slightly different (to my understanding anyway).

e.g. to report an error I do the following with the pipeline name as a prefix:

```nextflow
if (params.shortread_qc_includeunmerged && !params.shortread_qc_mergepairs) error("ERROR: [nf-core/taxprofiler] cannot include unmerged reads when merging is not turned on. Please specify --shortread_qc_mergepairs")
```
https://github.com/nf-core/taxprofiler/blob/45ce748534872c5105a5d3a94b83f47cd44f8826/workflows/taxprofiler.nf#L36

vs a warning

```nextflow
if (params.diamond_save_reads              ) log.warn "[nf-core/taxprofiler] DIAMOND only allows output of a single format. As --diamond_save_reads supplied, only aligned reads in SAM format will be produced, no taxonomic profiles will be available."
```
https://github.com/nf-core/taxprofiler/blob/45ce748534872c5105a5d3a94b83f47cd44f8826/workflows/taxprofiler.nf#L48

Where I have to manually specify `ERROR:` in when using `error()`, but the `log.warn()` autmatically puts `WARN` at the beginning. I guess there is a `log.error` but my understanding this doesn't cause the pipeline to exit.

## Suggest implementation 

Could be nice to have something a native `pipelineError("")` - thing the prefixes the workflow name at the beginning etc. 

It could also have a distinct colour from a Nextflow level error.

## More info

I initially brought this up for discussion with nf-core (in particular @ewels and @maxulysse) here: https://nfcore.slack.com/archives/C04QR0T3G3H/p1713533132261319 based on the suggestion from @Midnighter here: https://github.com/nf-core/taxprofiler/pull/60#discussion_r852845213.

All three may be able to provide better (technical) suggestions and/or benefits.

**Comment by ewels** [comment]
Optimum would be if Nextflow can determine where the `error` calls come from and figure out if it's coming from pipeline source code. Could then prefix with `manifest.name` if set.

**Comment by Midnighter** [comment]
These are two slightly different beasts. For the logger, it could be as simple as [modifying the layout](https://logback.qos.ch/manual/layouts.html) of messages. Although, we would then need two different loggers, one (or more) for core nextflow and another exposed to the pipeline that can be customized without affecting the others.

I don't know for sure, but I'm guessing that errors do not use logging facilities and thus need to be customized separately. Or indeed define a simple function/class method that performs some message layouting and then exits with an error code.

**Comment by jfy133** [comment]
Yes, sorry to be clear: I'm aware that the logging and error message are generated in two separate ways. 

I meant more that in this case, given I'm requesting that pipeline developers can define their own errors or warnings, these should follow a common structure for familiarity for the pipeline user.

**Comment by mahesh-panchal** [comment]
Adding my comment here:
https://github.com/nf-core/fetchngs/pull/309#discussion_r1587551622

One of my issues was that something ends up in the channel too if `ifEmpty` is used. Although this might well be solved with DSL3.

**Comment by Gullumluvl** [comment]
Hello,

I also have questions about logging, started on Slack at https://nextflow.slack.com/archives/C02T98A23U7/p1748442135074119, but I'll reorganize my thoughts here.

### Logging

First, as a pipeline developer, I would like to know how to achieve the following typical logging features (maybe these are points that need a convenient API, or if it's already doable, that need to be put clearly in the future doc?): 

* I would like to write logs in:
  + a specific file (not `.nextflow.log`), customizable.
  + to the screen
* Choose a different filter level for logging to screen (e.g. severity ≥ warning) than to file (≥ info or debug).
* modifying the layout as said above, to identify from which subworkflow/module a message comes from.

I also see challenges specific to the concurrent nature of workflows:

* how to log the content of a channel or send log messages that depend on a channel?

About this last point, it might be nice to provide a single obvious way to achieve this; currently I can use the `view` or `dump` operators, but how about a `log` operator?

### Errors, from channels

Similarly: could you have an *operator* that triggers an error? I tried to put `error()` calls inside `subscribe` operator, but it was behaving strangely (like I had `subscribe {println "some message"; error("error message"); }` but the first message was not showing up. Haven't thoroughly tested this though).

Then such operator should have a configuration/option to control how the workflow exits:

- immediately, like when you call `error` at the workflow level;
- or like a process, so it would follow the configuration for process failure (such as waiting for running tasks to complete)

Now even nicer would be that this operator also accepts a condition as an argument, to decide to fail or not.

In summary, that operator could be something like:

```nextflow
fail( {condition}, message: {"Problem with $it"}, level: "workflow")
```

with `level` being one of `"workflow"` or `"channel"` ?

Sorry, a bit long, hope this is relevant!
